File: NoveltySliceMediaItem.py
from reaper_python import *
import os
import sys
import shutil
import tempfile as tf
import subprocess

def sampstos(samples, sample_rate):
  return float((samples / sample_rate))

cancel, _, _, captions, user_input, _ = RPR_GetUserInputs('Novelty Slice Parameters', 5, 'feature,threshold,kernelsize,filtersize,fftsettings', '0,0.5,3,1,1024 512 1024', 512)
# parse user input
if cancel != 0:
  params = user_input.split(',')
  
  feature = params[0]
  threshold = params[1]
  kernelsize = params[2]
  filtersize = params[3]
  fftsettings = params[4]
  
  temp_dir = tf.mkdtemp()
  temp_idx = os.path.join(temp_dir, 'fluid_novelty_slice_reaper.wav')

  # Each user MUST point this to their folder containing FluCoMa CLI executables
  cli_path = '/Users/jamesbradbury/dev/bin'
  # Then we form some calls to the tools that will live in that folder
  index_extractor_exe = os.path.join(cli_path, 'index_extractor')
  novelty_slice_exe = os.path.join(cli_path, 'noveltyslice')

  # Get info for item in REAPER
  item = RPR_GetSelectedMediaItem(0, 0)
  take = RPR_GetActiveTake(item)
  src = RPR_GetMediaItemTake_Source(take)
  sr = RPR_GetMediaSourceSampleRate(src)
  _, full_path, _ = RPR_GetMediaSourceFileName(src, '', 1024)
  item_pos = RPR_GetMediaItemInfo_Value(item, "D_POSITION")

  novelty_slice = subprocess.Popen([
    novelty_slice_exe, 
    '-source', full_path,
    '-indices', temp_idx,
    '-feature', feature,
    '-kernelsize', kernelsize,
    '-threshold', threshold,
    '-filtersize', filtersize,
    '-fftsettings', fftsettings[0], fftsettings[1], fftsettings[2]])
  novelty_slice.wait()

  idx_extract = subprocess.run([index_extractor_exe, temp_idx], stdout=subprocess.PIPE, stderr=subprocess.PIPE).stdout.decode('utf-8')

  slice_points = idx_extract.split()

  for i in range(1, len(slice_points)):
    slice_pos = sampstos(
      float(slice_points[i]),
      sr
    ) 
    item = RPR_SplitMediaItem(item, item_pos + slice_pos)
  RPR_UpdateArrange()

  # get track
  #track = RPR_GetSelectedTrack(0, 0)
  #track_number = RPR_GetMediaTrackInfo_Value(track, "IP_TRACKNUMBER")
  #new_track = RPR_InsertTrackAtIndex(int(track_number), True)

  shutil.rmtree(temp_dir)

File: NoveltySliceMediaItem.lua
-- First we require the utilities found in FluidUtils.lua
-- This is a horrible workaround specific to REAPER
-- It forms a manual path to the path of this script
-- It looks relative to this script and uses dofile to import
local info = debug.getinfo(1,'S');
script_path = info.source:match[[^@?(.*[\/])[^\/]-$]]
dofile(script_path .. "FluidUtils.lua")

------------------------------------------------------------------------------------
--   Each user MUST point this to their folder containing FluCoMa CLI executables --
sanity_check()
local cli_path = get_fluid_path()
--   Then we form some calls to the tools that will live in that folder --
local ie_exe = cli_path .. '/index_extractor '
local ns_exe = cli_path .. '/noveltyslice '
------------------------------------------------------------------------------------

local num_selected_items = reaper.CountSelectedMediaItems(0)
if cancel ~= false and num_selected_items > 0 then
    local cancel, user_inputs = reaper.GetUserInputs("Novelty Slice Parameters", 5, "feature,threshold,kernelsize,filtersize,fftsettings", "0,0.5,3,1,1024 512 1024")
    local items = {}
    for x=0, num_selected_items-1 do
      table.insert(items, reaper.GetSelectedMediaItem(0, x))
    end
    for k=1, #items do
        local item = items[k]
        --local proj_path = reaper.GetProjectPath(0, "")
        --proj_path = doublequote(proj_path)
        tmp_file = os.tmpname() .. ".wav"
        temp_idx = doublequote(tmp_file)
      
        local params = commasplit(user_inputs)
        local feature = params[1]
        local threshold = params[2]
        local kernelsize = params[3]
        local filtersize = params[4]
        local fftsettings = params[5]
        
        -- Get info for item in REAPER
        local take = reaper.GetActiveTake(item)
        local src = reaper.GetMediaItemTake_Source(take)
        local sr = reaper.GetMediaSourceSampleRate(src)
        local full_path = reaper.GetMediaSourceFileName(src, '')
        full_path = doublequote(full_path)
        local item_pos = reaper.GetMediaItemInfo_Value(item, "D_POSITION")
        local ns_cmd = ns_exe .. " -source " .. full_path .. " -indices " .. temp_idx .. " -feature " .. feature .. " -kernelsize " .. kernelsize .. " -threshold " .. threshold .. " -filtersize " .. filtersize .. " -fftsettings " .. fftsettings
        local ie_cmd = ie_exe .. " " .. temp_idx

        os.execute(ns_cmd)
        local slice_points_string = capture(ie_cmd, false)
        local slice_points = spacesplit(slice_points_string)
        for i=2, #slice_points do
            local t_conversion = tonumber(slice_points[i])
            local slice_pos = sampstos(t_conversion, sr)
            item = reaper.SplitMediaItem(item, item_pos + slice_pos)
        end

        local kill_cmd = "rm -rf " .. temp_idx
        os.execute(kill_cmd)
    end
    reaper.UpdateArrange()
end
